### PR TITLE
Ensemble Rom

### LINKED ISSUE #1722: [TASK] Add Ensemble ROM Capability
--------
Issue Description
--------

**Is your feature request related to a problem? Please describe.**
The goal of ensemble ROMs (estimators) is to combine the predictions of several base ROMs (estimators) built with a given learning algorithm in order to improve generalizability / robustness over a single ROM.

Two families of ensemble ROMs are usually distinguished:

In averaging methods, the driving principle is to build several estimators independently and then to average their predictions. On average, the combined estimator is usually better than any of the single base estimator because its variance is reduced.

Examples: Bagging methods, Forests of randomized trees, …

By contrast, in boosting methods, base estimators are built sequentially and one tries to reduce the bias of the combined estimator. The motivation is to combine several weak models to produce a powerful ensemble.

Examples: AdaBoost, Gradient Tree Boosting, …

**Describe the solution you'd like**

**Describe alternatives you've considered**

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by wangcj05**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
close #1722 

The goal of ensemble ROMs (estimators) is to combine the predictions of several base ROMs (estimators) built with a given learning algorithm in order to improve generalizability / robustness over a single ROM.

Two families of ensemble ROMs are usually distinguished:

In averaging methods, the driving principle is to build several estimators independently and then to average their predictions. On average, the combined estimator is usually better than any of the single base estimator because its variance is reduced.

Examples: Bagging methods, Forests of randomized trees, …

By contrast, in boosting methods, base estimators are built sequentially and one tries to reduce the bias of the combined estimator. The motivation is to combine several weak models to produce a powerful ensemble.

Examples: AdaBoost, Gradient Tree Boosting, …

##### What are the significant changes in functionality due to this change request?

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: doc/user_manual/generated/generateRomDoc.py [hunk 1]
```diff
@@ -748,9 +748,12 @@
               'Clusters',
               'Interpolated']
 validDNNRom = ['KerasMLPClassifier',
+              'KerasMLPRegression',
               'KerasConvNetClassifier',
               'KerasLSTMClassifier',
-              'KerasLSTMRegression']
+              'KerasLSTMRegression',
+              'KerasLSTMRegression'
```
Comment (PaulTalbot-INL) [inline doc/user_manual/generated/generateRomDoc.py:755]: Is this the same ROM listed twice?

File: framework/Models/PostProcessors/LimitSurface.py [hunk 2]
```diff
@@ -131,7 +131,6 @@ def _initializeLSpp(self, runInfo, inputs, initDict):
       self.ROM.initializeModel(settings)
     else:
       self.ROM = self.assemblerDict['ROM'][0][3]
-    self.ROM.reset()
```
[the remark this answers is not in the document]
Comment (wangcj05) [reply framework/Models/PostProcessors/LimitSurface.py:134]: I just do not see the value for the reset here, it seems to me it is at the initialization stage, the ROM is initialized during this stage, why do we need to reset it?

File: doc/user_manual/generated/generateRomDoc.py (hunk 1 above)
Comment (wangcj05) [reply doc/user_manual/generated/generateRomDoc.py:755]: Yes, I will remove it.

File: framework/Models/PostProcessors/LimitSurface.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (PaulTalbot-INL) [reply framework/Models/PostProcessors/LimitSurface.py:134]: Hm, I don't disagree. I wonder why it was added in the first place.

**Comment by wangcj05** [comment]
@mandd Do you have time to review this PR?

File: framework/SupervisedLearning/ScikitLearn/Ensemble/__init__.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+  The ScikitLearn folder includes different types of Scikitlearn tree based ROMs
+  that are available via RAVEN
+"""
```
Comment (mandd) [inline framework/SupervisedLearning/ScikitLearn/Ensemble/__init__.py:17]: no imports here?

File: framework/SupervisedLearning/ScikitLearn/ScikitLearnBase.py [hunk 4]
```diff
@@ -74,17 +75,41 @@ def initializeModel(self, settings):
       @ In, settings, dict, the dictionary containin the parameters/settings to instanciate the model
       @ Out, None
     """
-    settings = self.updateSettings(settings)
-    self.settings = settings
-    self.model.set_params(**settings)
+    if self.settings is None:
+      self.settings = settings
```
Comment (mandd) [inline framework/SupervisedLearning/ScikitLearn/ScikitLearnBase.py:79]: should we set self.settings = settings only when settings is None?

File: doc/user_manual/generated/generateRomDoc.py (hunk 1 above)
Comment (wangcj05) [reply doc/user_manual/generated/generateRomDoc.py:754]: removed

File: framework/SupervisedLearning/ScikitLearn/Ensemble/__init__.py (hunk 3 above)
Comment (wangcj05) [reply framework/SupervisedLearning/ScikitLearn/Ensemble/__init__.py:17]: yes, no import is needed here.

File: framework/SupervisedLearning/ScikitLearn/ScikitLearnBase.py (hunk 4 above)
Comment (wangcj05) [reply framework/SupervisedLearning/ScikitLearn/ScikitLearnBase.py:79]: I think so, self.settings is the initial settings for the model, which is only initialized when we parse the input XML. It is can be used to reset the model. That's why I added a 'if' check here.

**Review by mandd — APPROVED** [review]

**Comment by mandd** [comment]
PR is green to me. Merging

**Pull request merged**